### PR TITLE
[feature/dns] dns resource lifecycle

### PR DESCRIPTION
**Pull request by carterkozak**
==COMMIT_MSG==
dns resource lifecycle
==COMMIT_MSG==

### CODE REVIEW

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 1]
```diff
@@ -134,6 +173,12 @@ default DialogueDnsResolver dnsResolver() {
         }
 
         Optional<ExecutorService> blockingExecutor();
+
+        static Builder builder() {
+            return new Builder();
+        }
+
+        class Builder extends ImmutableReloadingParams.Builder {}
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: This part should be reverted, unnecessary diff

File: dialogue-clients/src/test/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorkerTest.java [hunk 2]
```diff
@@ -144,7 +144,8 @@ public void testResolvedAddressesChangesAfterStartup() throws Exception {
                 .build();
         SettableRefreshable<ServicesConfigBlock> inputRefreshable = Refreshable.create(initialState);
         SettableRefreshable<ServicesConfigBlockWithResolvedHosts> receiverRefreshable = Refreshable.create(null);
-        DialogueDnsResolutionWorker worker = new DialogueDnsResolutionWorker(resolver, receiverRefreshable::update);
+        DialogueDnsResolutionWorker worker =
+                new DialogueDnsResolutionWorker(resolver, Duration.ofSeconds(5), receiverRefreshable);
```
Comment (carterkozak) [inline dialogue-clients/src/test/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorkerTest.java:148]: I think we may want to externally trigger state changes in this tests dns resolver, rather than leveraging interactions to change the state.

I don't want to do that in this PR though, since it's already complex enough

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 3]
```diff
@@ -170,27 +229,37 @@ public PerHostClientFactory perHost(String serviceName) {
         Preconditions.checkNotNull(serviceName, "serviceName");
         String channelName = ChannelNames.reloading(serviceName, params);
 
-        Refreshable<List<DialogueChannel>> perHostDialogueChannels = params.scb()
+        Refreshable<List<DialogueChannel>> perHostDialogueChannels = getServiceRefreshable(serviceName)
```
Comment (carterkozak) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: Updated to handle `perHost` here, the only remaining caller into the legacy `getNonReloadingChannel` over is the non-reloading method which takes a fully realized ServiceConfiguration as input. We can handle that in a separate PR, as well as the `getNonReloading(Class<T> clazz, ClientConfiguration clientConf)` overload

File: dialogue-clients/src/test/java/com/palantir/dialogue/otherpackage/DialogueClientsTest.java [hunk 4]
```diff
@@ -169,9 +165,9 @@ void perHost_attributes_concurrencylimiter_metrics_to_the_right_host() {
                         .putServices(
                                 "my-service",
                                 PartialServiceConfiguration.builder()
-                                        .addUris("https://my-service-0.fake.palantir.com")
-                                        .addUris("https://my-service-1.fake.palantir.com")
-                                        .addUris("https://my-service-2.fake.palantir.com")
+                                        .addUris("https://127.0.0.1/0")
+                                        .addUris("https://127.0.0.1/1")
+                                        .addUris("https://127.0.0.1/2")
```
Comment (carterkozak) [inline dialogue-clients/src/test/java/com/palantir/dialogue/otherpackage/DialogueClientsTest.java:170]: Updated because the manifestation is slightly different -- this way we allow the actual client to progress as expected.
Without using localhost here, the dns resolution component reduces the target to no TargetUris as dns resolution fails.

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorker.java [hunk 5]
```diff
@@ -64,15 +70,23 @@ public void run() {
         while (!shutdownRequested) {
             try {
                 // check for updates to scb state first
-                Thread.sleep(5000L);
+                Thread.sleep(dnsRefreshInterval.toMillis());
                 try {
+                    if (receiver.get() == null) {
+                        shutdownRequested = true;
+                        log.info("Output refreshable has been garbage collected, no need to continue polling");
```
Comment (bjlaub) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorker.java:77]: should there be an `else` or a `break` here?

Comment (carterkozak) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorker.java:77]: Ya, we should probably add an `else` here to avoid unnecessary work. `doUpdate` must always handle the not-present case either way, as the gc may run between doing a pre-check and later on.

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java [hunk 6]
```diff
@@ -382,105 +456,106 @@ private ClientConfiguration hydrate(ClientConfiguration configuration) {
         return configuration;
     }
 
-    // TODO(ckozak): Each of the 'map(' blocks should be extracted to another method. I'm not doing that
-    // yet be cause it will make the diff much harder to review.
-    @SuppressWarnings("CyclomaticComplexity")
+    private Refreshable<InternalDialogueChannelConfiguration> getServiceRefreshable(String serviceName) {
```
Comment (bjlaub) [inline dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: minor nit: i'm not wild about this method name, perhaps something like `configurationForService` as this returns a configuration that can be used to create a channel for this service? not super picky though

**Review by bjlaub — DISMISSED** [review]

File: dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java (hunk 6 above)
Comment (carterkozak) [reply dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java]: 👍 One of the two hardest problems in CS
